Add message branding with types

odashevskii-plaid opened the pull request:
I've originally commented [here](https://github.com/timostamm/protobuf-ts/issues/58#issuecomment-906874300) on why this is important; in a nutshell, the existing layout when messages are fully decoupled from their types doesn't work in situations when you need to reflect on a message which came from elsewhere and its exact type is unknown.

Here's a stripped-down example:

```proto
// PROTO SOURCE

// Data classification levels
enum Classification {
  // Non-sensitive data concerning Big Corp systems or users
  CLASSIFICATION_NON_SENSITIVE = 0;
  // Sensitive data concerning either consumers or developers (users of Big Corp)
  CLASSIFICATION_USER_SENSITIVE = 1;
}

extend google.protobuf.FieldOptions {
  // specifies data classfication
  Classification classification = 50001;
}

message Client {
  string client_id = 1;
  string secret = 2 [(annotations.classification) = CLASSIFICATION_USER_SENSITIVE];
}
```

```typescript
// TS CODE

const client1 = Client.create({ ... });
const client2 = Client.create({ ... });

enterpriseGradeLogger.info("got something", {
  foo: "bar",
  clients: [client1, client2],
});
```

The second argument of `enterpriseGradeLogger` is an arbitrary object; recursive redaction is applied, and for any protobuf  messages, redaction is based on field annotations.

With this PR, redaction logic could check if `reflectionIsProtoMessage` returns true and proceed with extracting annotations if so.

timostamm reviewed (commented):
Thank you for this, @odashevskii-plaid. I've run into cases like your logger example several times already. I've added a [comment](https://github.com/timostamm/protobuf-ts/issues/58#issuecomment-907808998) about the type property and my thoughts on it. 

As far as I see, this could be a first step into this direction. I'm okay with it not being part of the generated interfaces for a start. I'm also okay with not caring about `IMessageType.is`, `isAssignable` and `equals` for now. But some points should be addressed before we can sneak this in.

odashevskii-plaid commented on packages/runtime/src/reflection-create.ts:
```diff
@@ -1,14 +1,16 @@
+    let msg: UnknownMessage = {
+        [REFLECTION_BRAND]: info,
+    };
     for (let field of info.fields) {
```
In fact, it's not overridden! The `Create` instance is instantiated [here](https://github.com/timostamm/protobuf-ts/blob/master/packages/plugin/src/code-gen/message-type-generator.ts#L49), but not used any further. You probably intended to add it to `classDecMembers` [here](https://github.com/timostamm/protobuf-ts/blob/master/packages/plugin/src/code-gen/message-type-generator.ts#L117-L119)...

timostamm replied:
Hah, good shout! And it looks like some regressions hide behind that. I'm getting an error TS1371 when running `make` in `packages/plugin/` when I enable the code. Let me look at that quickly.

odashevskii-plaid commented:
@timostamm thanks for the quick turnaround and detailed suggestions! I think I have addressed all your comments except adding `[MESSAGE_TYPE]: ...` to the generated code (but see my comment). I will try fixing `create` too.

odashevskii-plaid commented on packages/runtime/src/reflection-get-type.ts:
```diff
@@ -0,0 +1,25 @@
+
+    if (msgType == null) {
+        throw Error('reflectionGetType: msg is not a protobuf message');
+    }
```
Note: I had to make the argument type be `object`, otherwise it didn't compile.

timostamm commented:
Thanks so far, @odashevskii-plaid. `create` has been fixed in https://github.com/timostamm/protobuf-ts/commit/05348911be432c5f654e2a8462b0280a05c7ea69. 

For context:

https://github.com/timostamm/protobuf-ts/blob/05348911be432c5f654e2a8462b0280a05c7ea69/packages/plugin/src/message-type-extensions/create.ts#L73-L87

- during "compilation", we have a in-memory `MessageType` for the message we want to generate code for
- `defaultMessage` is created from this `MessageType` with `create()`
- `typescriptLiteralFromValue(defaultMessage)` turns the object into an AST

I don't think it's possible to simply add `MESSAGE_TYPE` to defaultMessage because `typescriptLiteralFromValue` only understands primitive values. You will probably have to manipulate the AST , or generate a new line of code that does `defaultMessage.MESSAGE_TYPE = ...`.

odashevskii-plaid commented:
@timostamm thanks for the suggestions, I was able to figure out the Compiler API, see 483882e. The final `create` body looks like this:

```typescript
    // packages/test-generated/ts-out/google/type/color.ts
    create(value?: PartialMessage<Color>): Color {
        const message = { red: 0, green: 0, blue: 0 };
        (message as unknown as MessageTypeContainer<Color>)[MESSAGE_TYPE] = this;
        if (value !== undefined)
            reflectionMergePartial<Color>(this, message, value);
        return message;
    }
```

timostamm commented on packages/runtime/src/reflection-contains-message-type.ts:
```diff
@@ -0,0 +1,22 @@
+/**
+ * Check if the provided object is a proto message.
+ */
+export function containsMessageType<T extends object>(msg: object): msg is MessageTypeContainer<T> {
+    return (msg as MessageTypeContainer<T>)[MESSAGE_TYPE] != null;
+}
+
```
The typing I suggested originally was incomplete. This will work:

```suggestion
export function containsMessageType<T extends object>(msg: T): msg is (T & MessageTypeContainer<T>) {
    return (msg as MessageTypeContainer<T>)[MESSAGE_TYPE] != null;
}

```

It let's the compiler know that `msg` has the `MESSAGE_TYPE` property:

```ts
if (containsMessageType(msg)) {
     msg[MESSAGE_TYPE].toJson(msg)
}
```

It will still let you access regular properties of `msg` and also works on `UnknownMessage`.

odashevskii-plaid replied:
👍

timostamm commented on packages/runtime/src/reflection-contains-message-type.ts:
```diff
@@ -0,0 +1,22 @@
+    return (msg as MessageTypeContainer<T>)[MESSAGE_TYPE] != null;
+}
+
+/**
+ * Get a message type from a proto message.
+ */
+export function getMessageType<T extends object>(msg: MessageTypeContainer<T>): IMessageType<T> {
+    return msg[MESSAGE_TYPE];
+}
```
```suggestion
```

With the typing change above, I don't see any benefit from having this function.

You could easily express this with

```ts
import {containsMessageType, MESSAGE_TYPE} from "@protobuf-ts/runtime";
const type = containsMessageType(msg) ? msg[MESSAGE_TYPE] : undefined;
```

Or

```ts
if (containsMessageType(msg)) {
   msg[MESSAGE_TYPE].toJson(msg)
} else {
   throw "cannot redact"
}
```

Or 

````ts
assert(containsMessageType(msg), "cannot redact");
msg[MESSAGE_TYPE].toJson(msg)
```

Would you agree?

timostamm commented on packages/runtime/spec/reflection-contains-message-type.spec.ts:
```diff
@@ -0,0 +1,42 @@
+import {MessageType, ScalarType, containsMessageType, getMessageType} from '../src';
+
+describe('containsMessageType', () => {
+    interface MyMessage {
+        stringField: string;
+    }
+
+    const MyMessage: MessageType<MyMessage> = new MessageType<MyMessage>('.test.MyMessage', [
+        {no: 1, name: 'string_field', kind: "scalar", T: ScalarType.STRING},
+    ]);
+
+    it('allows to extract type after .create', () => {
+        const msg = MyMessage.create({
+            stringField: "hello world",
+        });
+
+        if (containsMessageType(msg)) {
+            expect(getMessageType(msg)).toEqual(MyMessage);
+        } else {
+            fail("containsMessageType() must return true");
+        }
+    });
+
+    it('allows to extract type after .clone', () => {
+        const msg = MyMessage.create({
+            stringField: "hello world",
+        });
+
+        const cloned = MyMessage.clone(msg);
+        if (containsMessageType(cloned)) {
+            expect(getMessageType(cloned)).toEqual(MyMessage);
```
With the typing change suggested below:

```suggestion
import {MessageType, ScalarType, containsMessageType, MESSAGE_TYPE} from '../src';

describe('containsMessageType', () => {
    interface MyMessage {
        stringField: string;
    }

    const MyMessage: MessageType<MyMessage> = new MessageType<MyMessage>('.test.MyMessage', [
        {no: 1, name: 'string_field', kind: "scalar", T: ScalarType.STRING},
    ]);

    it('allows to extract type after .create', () => {
        const msg = MyMessage.create({
            stringField: "hello world",
        });

        if (containsMessageType(msg)) {
            expect(msg[MESSAGE_TYPE]).toEqual(MyMessage);
        } else {
            fail("containsMessageType() must return true");
        }
    });

    it('allows to extract type after .clone', () => {
        const msg = MyMessage.create({
            stringField: "hello world",
        });

        const cloned = MyMessage.clone(msg);
        if (containsMessageType(cloned)) {
            expect(cloned[MESSAGE_TYPE]).toEqual(MyMessage);
```

timostamm reviewed (commented):
Apologies for suggesting incomplete typing for `containsMessageType` earlier, @odashevskii-plaid. Have added a suggestion to address. If you could implement it, this is ready to merge.

odashevskii-plaid commented:
@timostamm the new typing is cool; I've removed `getMessageType`...

jcready commented:
@timostamm I understand the use-case here, but IMO this really clutters up the objects returned from `.create()` and `.fromBinary()`. Is it possible to make this some kind of flag/option during codegen to disable this for consumers that do not need to perform reflection on in-memory objects?

I realize this wasn't a breaking change, but trying to update our codebase from 2.0.2 to 2.0.3 has been made difficult because we have some unit tests that are doing things like `expect(decodedMessage).toEqual(expectedOutput)` which are no longer passing. We either need to loosen up our expectations (e.g. replace `.toEqual()` with something like `.toMatchObject()`) or we need to `const expectedOutput = IMessageType.create({ ...previouslyValidPlainJSObject });`. Either way it's difficult for me to explain to my colleagues why a patch version bump requires their tests to be updated.

timostamm commented:
If tests break, this is a breaking change in my mind. But why do they break? 

Symbol properties are unlike regular properties (not listed in `Object.keys` for example) and this test (just added) passes:

https://github.com/timostamm/protobuf-ts/blob/1f6771f55712180effb7817211baef53744f674a/packages/runtime/spec/message-type.spec.ts#L86-L100

It is using Jasmine 3.5. Could you share details about your setup?

jcready commented:
We're using Jest v26. Docs for `.toEqual()` here: https://jestjs.io/docs/26.x/expect#toequalvalue

I tracked down Jest's implementation of `.toEqual()` to here: https://github.com/facebook/jest/blob/bc50e7f360ab1845abbaa0b3ad788caead0d3174/packages/expect/src/jasmineUtils.ts#L29-L38, but it does look like it's comparing enumerable symbols: https://github.com/facebook/jest/blob/bc50e7f360ab1845abbaa0b3ad788caead0d3174/packages/expect/src/jasmineUtils.ts#L208-L214

jcready commented:
Does this symbol need to be enumerable? It seems like @odashevskii-plaid's use-case would still be met if it was non-enumerable, but perhaps his example was simplified and he would still need this to work after stuff like object spread.

timostamm commented:
I disagree with Jest's choice and believe that symbols should only be compared with `toStrictEqual`. But Jest is _very_ widely used. Release [v2.0.4-alpha.0](https://github.com/timostamm/protobuf-ts/releases/tag/v2.0.4-alpha.0) in the `next` channel makes the property non-enumerable. 

@odashevskii-plaid, you can still access the `MESSAGE_TYPE` like before. But if you copy a message with `{...msg}`, the property will be lost - please use `MyMessage.clone()` instead.

@jcready, it would be great if you could confirm the fix with Jest.

jcready commented:
I can confirm that making the property non-enumerable fixes the issue. Thank you for the incredibly quick turnaround. I hope no one else is broken by making it non-enumerable.

timostamm commented:
> I can confirm that making the property non-enumerable fixes the issue. 

Thanks! Released in [v2.0.4](https://github.com/timostamm/protobuf-ts/releases/tag/v2.0.4).
